### PR TITLE
fix grouping by clustered by column on partitioned table

### PR DESCRIPTION
**Pull request by mfelsche**

### CODE REVIEW

File: sql/src/main/java/io/crate/planner/consumer/ReduceOnCollectorGroupByConsumer.java [hunk 1]
```diff
@@ -76,17 +72,17 @@ public Context(ConsumerContext context) {
...
-        }
-
         @Override
         public AnalyzedRelation visitQueriedTable(QueriedTable table, Context context) {
             if (table.querySpec().groupBy() == null) {
                 return table;
             }
+
+            // no row authority on shards for partitioned tables
+            if (!table.tableRelation().tableInfo().partitionedBy().isEmpty()) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/consumer/ReduceOnCollectorGroupByConsumer.java]: TableInfo.isPartitioned()

File: sql/src/main/java/io/crate/planner/consumer/ReduceOnCollectorGroupByConsumer.java [hunk 2]
```diff
@@ -52,11 +52,7 @@
 
 public class ReduceOnCollectorGroupByConsumer implements Consumer {
 
-    private final Visitor visitor;
-
-    public ReduceOnCollectorGroupByConsumer(AnalysisMetaData analysisMetaData) {
-        visitor = new Visitor(analysisMetaData);
-    }
+    private final Visitor visitor = new Visitor();
```
Comment (dobe) [inline sql/src/main/java/io/crate/planner/consumer/ReduceOnCollectorGroupByConsumer.java]: could be even static 
and upercase then :)

File: sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java [hunk 3]
```diff
@@ -1019,4 +1019,35 @@ public void groupByStatsAggregatesGlobal() throws Exception {
...
     }
+
+    @Test
+    public void testGroupByOnClusteredByColumnPartitioned() throws Exception {
+        execute("CREATE TABLE tickets ( " +
+                "  ticket_id long, " +
+                "  tenant_id integer, " +
+                "  creator_id integer, " +
+                "  created timestamp, " +
+                "  created_month string )" +
```
Comment (dobe) [inline sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java]: remove stuff from schema not related to the test

File: sql/src/main/java/io/crate/planner/consumer/QueryThenFetchConsumer.java [hunk 4]
```diff
@@ -35,11 +34,7 @@
 
 public class QueryThenFetchConsumer implements Consumer {
 
-    private final Visitor visitor;
-
-    public QueryThenFetchConsumer(AnalysisMetaData analysisMetaData) {
-        visitor = new Visitor(analysisMetaData);
-    }
+    private static final Visitor visitor = new Visitor();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/consumer/QueryThenFetchConsumer.java]: upper case ?

File: sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java [hunk 5]
```diff
@@ -55,12 +55,7 @@
 
 public class QueryAndFetchConsumer implements Consumer {
 
-    private final Visitor visitor;
-
-    public QueryAndFetchConsumer(AnalysisMetaData analysisMetaData){
-        visitor = new Visitor(analysisMetaData);
-    }
-
+    private static final Visitor visitor = new Visitor();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java]: upper case

File: sql/src/main/java/io/crate/planner/consumer/NonDistributedGroupByConsumer.java [hunk 6]
```diff
@@ -45,11 +45,7 @@
 
 public class NonDistributedGroupByConsumer implements Consumer {
 
-    private final Visitor visitor;
-
-    public NonDistributedGroupByConsumer(AnalysisMetaData analysisMetaData) {
-        visitor = new Visitor(analysisMetaData);
-    }
+    private static final Visitor visitor = new Visitor();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/consumer/NonDistributedGroupByConsumer.java]: here too

File: sql/src/main/java/io/crate/planner/node/dql/GroupByConsumer.java [hunk 7]
```diff
@@ -57,7 +57,7 @@ public static boolean groupedByClusteredColumnOrPrimaryKeys(TableRelation tableR
                     .equals(tableRelation.tableInfo().clusteredBy()));
     }
 
-    private static boolean groupedByPrimaryKeys(TableRelation tableRelation, List<Symbol> groupBy) {
+    public static boolean groupedByPrimaryKeys(TableRelation tableRelation, List<Symbol> groupBy) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/node/dql/GroupByConsumer.java]: why is it now public? It is not used anywhere else?

File: sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java [hunk 8]
```diff
@@ -1015,8 +1015,37 @@ public void groupByStatsAggregatesGlobal() throws Exception {
...
+    @Test
+    public void testGroupByOnClusteredByColumnPartitioned() throws Exception {
+        execute("CREATE TABLE tickets ( " +
+                "  ticket_id long, " +
+                "  tenant_id integer, " +
+                "  created_month string )" +
+                "PARTITIONED BY (created_month) " +
+                "CLUSTERED BY (tenant_id) " +
+                "WITH (number_of_replicas=0)");
+        ensureGreen();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java]: isn't ensureYellow enough here?

File: sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java [hunk 9]
```diff
@@ -1015,8 +1015,37 @@ public void groupByStatsAggregatesGlobal() throws Exception {
...
+        execute("insert into tickets (ticket_id, tenant_id, created_month) values (?, ?, ?)",
+                new Object[][]{
+                        {1L, 42, 1425168000000L},
+                        {2L, 43, 0},
+                        {3L, 44, 1425168000000L},
+                        {4L, 45, 499651200000L},
+                        {5L, 42, 0L},
+                });
+        ensureGreen();
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java]: same here

File: sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java [hunk 10]
```diff
@@ -1015,8 +1016,38 @@ public void groupByStatsAggregatesGlobal() throws Exception {
...
         assertThat((Double) response.rows()[0][2], closeTo(47.84415001097868d, 0.0000001));
-        assertThat((Short)response.rows()[0][3], is((short)112));
-        assertThat((Double)response.rows()[0][4], is(1090.6875d));
-        assertThat((Double)response.rows()[0][5], is(33.025558284456d));
+        assertThat((Short) response.rows()[0][3], is((short) 112));
+        assertThat((Double) response.rows()[0][4], is(1090.6875d));
+        assertThat((Double) response.rows()[0][5], is(33.025558284456d));
+    }
+
+    @Repeat(iterations=100)
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/GroupByAggregateTest.java]: Repeat ?

**Comment by mfussenegger** [comment]
ack

**Pull request merged**